lib/DependencyProvider.js

```diff
@@ -0,0 +1,317 @@
+  } catch (_) {
+    const remoteUrl = remoteElmJsonUrl(pkg, version);
+    const elmJson = request('GET', remoteUrl).getBody('utf8'); // need utf8 to convert from gunzipped buffer
+    const cachePath = cacheElmJsonPath(pkg, version);
+    const parentDir = path.dirname(cachePath);
+    fs.mkdirSync(parentDir, { recursive: true });
+    fs.writeFileSync(cachePath, elmJson);
+    return elmJson;
+  }
+}
+
+function fetchElmJsonOffline(
+  pkg /*: string */,
+  version /*: string */
+) /*: string */ {
+  // console.log('Fetching: ' + pkg + ' @ ' + version);
+  try {
+    return fs.readFileSync(homeElmJsonPath(pkg, version), 'utf8');
+  } catch (_) {
+    try {
+      return fs.readFileSync(cacheElmJsonPath(pkg, version), 'utf8');
+    } catch (_) {
+      throw `Offline mode, so we fail instead of doing a remote request.`;
+    }
+  }
+}
+
+function updateOnlineVersionsCache() /*: void */ {
+  const pubgrubHome = path.join(elmHome(), 'pubgrub');
+  fs.mkdirSync(pubgrubHome, { recursive: true });
+  const cachePath = path.join(pubgrubHome, 'versions_cache.json');
+  const remotePackagesUrl = 'https://package.elm-lang.org/all-packages';
+  if (onlineVersionsCache.size === 0) {
+    try {
+      // Read from disk what is already cached, and complete with a request to the package server.
+      const cache = JSON.parse(fs.readFileSync(cachePath, 'utf8'));
+      onlineVersionsCache = parseOnlineVersions(cache);
+      updateCacheWithRequestSince(cachePath, remotePackagesUrl);
+    } catch (_) {
+      // The cache file does not exist, let's download it all.
+      updateCacheFromScratch(cachePath, remotePackagesUrl);
+    }
+  } else {
+    // The cache is not empty, we just need to update it.
+    updateCacheWithRequestSince(cachePath, remotePackagesUrl);
+  }
+}
+
+// Reset the cache of existing versions from scratch
+// with a request to the package server.
+function updateCacheFromScratch(
+  cachePath /*: string */,
+  remotePackagesUrl /*: string */
+) /*: void */ {
+  const onlineVersionsJson = request('GET', remotePackagesUrl).getBody('utf8');
+  fs.writeFileSync(cachePath, onlineVersionsJson);
+  const onlineVersions = JSON.parse(onlineVersionsJson);
+  onlineVersionsCache = parseOnlineVersions(onlineVersions);
+}
+
+// Update the cache with a request to the package server.
+function updateCacheWithRequestSince(
+  cachePath /*: string */,
+  remotePackagesUrl /*: string */
+) /*: void */ {
+  // Count existing versions.
+  let versionsCount = 0;
+  for (const versions of onlineVersionsCache.values()) {
+    versionsCount += versions.length;
+  }
+
+  // Complete cache with a remote call to the package server.
+  const remoteUrl = remotePackagesUrl + '/since/' + (versionsCount - 1); // -1 to check if no package was deleted.
+  const newVersions = JSON.parse(request('GET', remoteUrl).getBody('utf8'));
+  if (newVersions.length === 0) {
+    // Reload from scratch since it means at least one package was deleted from the registry.
```

Comment:
what happens if a package is deleted and then a different package is immediately uploaded?